### PR TITLE
adds add mxnet

### LINKED ISSUE #1836: readd mxnet
i have removed mxnet from the master.

reason:  it broke travis. 

also: the package is not an CRAN. it is currently unclear how we can integrate such a learner then in the official release. i do hope it is simple, but we need to define and doc this. 

lets resolve this soon as many peops will like this new learner.

### PR DESCRIPTION
**Pull request by berndbischl**
resolves #1836

### CODE REVIEW

**Comment by ja-thomas** [comment]
@bklppr I merged in your fork into this branch, let's continue all communication here. About implementation here.

File: R/RLearner_classif_mxff.R [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+        values = c("max", "avg", "sum"),
+        requires = quote(conv.layer2 == TRUE)),
+      makeDiscreteLearnerParam(id = "pool.type3", default = "max",
+        values = c("max", "avg", "sum"),
+        requires = quote(conv.layer3 == TRUE)),
+      # other hyperparameters
+      makeNumericLearnerParam(id = "validation.set"),
+      makeIntegerLearnerParam(id = "early.stop.badsteps", lower = 1),
+      makeLogicalLearnerParam(id = "early.stop.maximize", default = TRUE),
+      makeNumericVectorLearnerParam(id = "dropout", lower = 0, upper = 1 - 1e-7),
```
Comment (ja-thomas) [inline R/RLearner_classif_mxff.R]: Hm, so dropout might have different lengths (depending on if I want separate dropout ratios in every layer). This is again a bit annoying to tune. 
Maybe have them as separate parameters per layer and have an additional dropout.global. The separate layers then `require(dropout.global == FALSE)`.

Comment (ja-thomas) [inline R/RLearner_classif_mxff.R]: and one `dropout.input` which does not requires dropout.global to be FALSE

File: R/RLearner_classif_mxff.R [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+    `c(conv.dilate11, conv.dilate12)` and `c(conv.pad11, conv.pad12)` correspond to the parameters
+    of `mx.symbol.Convolution`. When a `Convolution` layer is constructed, a `Pooling` layer is
+    constructed with it automatically. Again sticking to the example of the first layer,
+    `c(pool.kernel11, pool.kernel12)`, `c(pool.stride11, pool.stride12)`, `c(pool.pad11, pool.pad12)`
+    and `c(pool.type11, pool.type12)` correspond to the parameters in `mx.symbol.Pooling`.
+    If only the first parameter of a pair is defined, the second parameter is set to the same value,
+    e.g. if `conv.kernel11` is defined and `conv.kernel12` is undefined, `conv.kernel12` is set to
+    the value of `conv.kernel11`.
+    When convolution is used, `conv.data.shape` needs to be specified, which is a vector giving the
+    dimensionality of the data (e.g. for MNIST `c(28, 28)`). Furthermore, `array.layout` is set to
```
Comment (ja-thomas) [inline R/RLearner_classif_mxff.R]: Does this also allow for three-dimensional inputs, e.g. images with color levels?

File: R/RLearner_classif_mxff.R [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+  # construct validation data
+  if (is.null(eval.data) & !is.null(validation.set)) {
+    eval.data = list()
+    n = dim(x)[1]
+    val.ind = sample(n, floor(n * validation.set))
+    eval.data$label = y[val.ind]
+    y = y[-val.ind]
+    eval.data$data = x[val.ind,]
+    x = x[-val.ind,]
+  }
```
Comment (ja-thomas) [inline R/RLearner_classif_mxff.R]: You might want to use `makeResampleDesc("Holdout", ...)` to make use of stratification

File: R/RLearner_classif_mxff.R [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+          dilate = conv.dilates[[i]], pad = conv.pads[[i]], num_filter = nums[i])
+        sym = do.call(mxnet::mx.symbol.Convolution, conv.inputs[!sapply(conv.inputs, is.null)])
+        sym = mxnet::mx.symbol.Activation(sym, act_type = act[i])
+        pool.inputs = list(data = sym, kernel = pool.kernels[[i]], pool.type = pool.types[[i]],
+          stride = pool.strides[[i]], pad = pool.pads[[i]])
+        sym = do.call(mxnet::mx.symbol.Pooling, pool.inputs[!sapply(pool.inputs, is.null)])
+      } else {
+        # construct fully connected layer
+        if (i > 1) {
+          if (convs[i - 1]) {
```
Comment (ja-thomas) [inline R/RLearner_classif_mxff.R]: please comment a bit more what is happening here

File: R/RLearner_classif_mxff.R [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+      c(.learner$par.vals$conv.data.shape, nrow(x)),
+      .learner$par.vals$conv.data.shape)
+    x = array(aperm(x), dim = dims)
+    array.layout = "colmajor"
+  }
+  p = predict(.model$learner.model, X = x, array.layout = array.layout)
+  if (.learner$predict.type == "response") {
+    p = apply(p, 2, function(i) {
+      w = which.max(i)
+      return(ifelse(length(w > 0), w, NaN))
```
Comment (ja-thomas) [inline R/RLearner_classif_mxff.R]: Please comment this a bit more, why do you return `NaN`

File: R/RLearner_classif_mxff.R [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+    y = y[-val.ind]
+    eval.data$data = x[val.ind,]
+    x = x[-val.ind,]
+  }
+
+  # if convolution is used, prepare the data dimensionality
+  if (conv.layer1) {
+    l = length(.learner$par.vals$conv.data.shape)
+    dims = switch(l,
+      c(.learner$par.vals$conv.data.shape, 1, 1, nrow(x)),
```
Comment (ja-thomas) [inline R/RLearner_classif_mxff.R]: please comment what kind of data each line refers to

File: R/RLearner_classif_mxff.R [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+        c(.learner$par.vals$conv.data.shape, 1, 1, nrow(eval.data$data)),
+        c(.learner$par.vals$conv.data.shape, 1, nrow(eval.data$data)),
+        c(.learner$par.vals$conv.data.shape, nrow(eval.data$data)),
+        .learner$par.vals$conv.data.shape)
+      eval.data$data = array(aperm(eval.data$data), dim = dims)
+    }
+  }
+
+  # early stopping
+  if (is.null(epoch.end.callback) & is.numeric(early.stop.badsteps)) {
```
Comment (ja-thomas) [inline R/RLearner_classif_mxff.R]: why are you checking `early.stop.badsteps` for numeric? It's an integer, I guess you want to check for NULL? Then use the correct check

File: R/RLearner_classif_mxff.R [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+      makeLogicalLearnerParam(id = "conv.layer2", default = FALSE,
+        requires = quote(layers > 1 && conv.layer1 == TRUE)),
+      makeLogicalLearnerParam(id = "conv.layer3", default = FALSE,
+        requires = quote(layers > 2 && conv.layer2 == TRUE)),
+      makeIntegerVectorLearnerParam(id = "conv.data.shape",
+        requires = quote(conv.layer1 == TRUE)),
+      makeIntegerLearnerParam(id = "conv.kernel11", lower = 1L,
+        requires = quote(conv.layer1 == TRUE)),
+      makeIntegerLearnerParam(id = "conv.kernel12", lower = 1L,
+        requires = quote(conv.layer1 == TRUE)),
```
Comment (bklppr) [inline R/RLearner_classif_mxff.R]: I think in principle you're right and that's the implementation I had originally. However, there is a problem with `integerVectorParams` when using them with `generateDesign` because it the vector parameters get split up into single integer parameters which can in turn not be used as input for the learner. We talked about this issue when it occurred :)
```
> test.ps = makeParamSet(makeIntegerVectorParam(id = "test", len = 2, lower = 0, upper = 10))
> generateDesign(n = 2, test.ps)

   test1 test2
1     7     2
2     2     8
```

File: R/RLearner_classif_mxff.R (hunk 2 above)
Comment (bklppr) [inline R/RLearner_classif_mxff.R]: Yes it should, `conv.data.shape` can have between length 1 and 4.

File: R/RLearner_classif_mxff.R (hunk 5 above)
Comment (bklppr) [inline R/RLearner_classif_mxff.R]: I've commented this in the code. I think `NA` is more appropriate here, please tell me what you think.

File: R/RLearner_classif_mxff.R (hunk 8 above)
Comment (ja-thomas) [inline R/RLearner_classif_mxff.R]: Yes, we talked about that and you're completely correct with the generate design function. Although in mlr tuning vector params **of fixed length** works. 

See for example with H2O:

```
library(mlr)
lrn = makeLearner('classif.h2o.deeplearning')
ps = makeParamSet(
   makeIntegerVectorParam('hidden', lower = c(1,1), upper = c(5,5), len = 2)

ctrl = makeTuneControlRandom(maxit = 1)
tuneParams(lrn, iris.task, hout, mmce, ps, ctrl)
 
```

**Comment by ja-thomas** [comment]
You still seem to have some missing prefixes I assume

```
* checking dependencies in R code ... WARNING
'::' or ':::' import not declared from: â€˜mxnetâ€™
```

**Comment by bklppr** [comment]
I could not find any missing prefixes. Running `ack [^:]mx\.` showed however that there are some `.Rhistory` and `.Rproj.user` files where the prefix is missing.

Also I found [this statement](https://stat.ethz.ch/pipermail/r-devel/2013-February/065965.htmlpoo0) from Hadley Wickham which might be relevant (?):

> Except the double colon method doesn't work (i.e. does not pass R CMD
> check) unless you also import the package, which means it's no longer
> just a suggestion - it must always be installed.
> 
> A simple test case (attached DESCRIPTION and R/test.r) yields this
> warning on R CMD check:
> 
> * checking for unstated dependencies in R code ... WARNING
> '::' or ':::' import not declared from: 'MASS'

File: R/RLearner_classif_mxff.R (hunk 8 above)
Comment (bklppr) [inline R/RLearner_classif_mxff.R]: Ah ok, I did not know this. This is very cool! 👍 
Especially that one can still generate a design first and then pass it to the tuneControl.

**Comment by ja-thomas** [comment]
You might want to add mxnet to the suggests. We still cannot test on travis I think, there are precompiled versions of mxnet for windows and Mac, but on Linux we still have to build it by hand (first mxnet core, then the R package)

**Comment by ja-thomas** [comment]
Btw. Your tests are producing warnings:

```
Warnings -----------------------------------------------------------------------
1. classif_mxff (@test_classif_mxff.R#39) - Auto detect layout of input matrix, use rowmajor..
```

**Comment by ja-thomas** [comment]
@bklppr I merged master into your branch so that the branch is building (Travis is fixed on the master) and also added mxnet to suggests, let's see how far Travis makes it this time :)

**Comment by berndbischl** [comment]
> You might want to add mxnet to the suggest

thats not a "might" but a "must"

**Comment by ja-thomas** [comment]
Highjacking this branch for a second.

I added regression and more importantly uncertainty estimation for it (according to http://mlg.eng.cam.ac.uk/yarin/blog_3d801aa532c1ce.html).

Feel free to ignore the regression stuff, but I need that now :)

**Comment by ja-thomas** [comment]
@bklppr We do currently not support Batch normalization, that is something that we should think about.

**Comment by bklppr** [comment]
@ja-thomas I added batch normalization, please tell me what you think.
Also, just to make sure, I implemented all your requested changes, right? Or did I miss something?

**Comment by bklppr** [comment]
@ja-thomas I added a minor update to the `predict` function of `regr.mxff`.
Also I added a simple unit test for `regr.mxff`. However, I was not able to use dropout for the learner as it uses the argument `mode` in `mx.symbol.Dropout` in the `train` function which does not seem to be a valid argument.

Regarding the planned merge of branch mxnet into master: To my knowledge I implemented all of your requested changes to `classif.mxff` so this part should be finished.
